app/src/main/java/com/github/se/eduverse/ui/search/UserProfileScreen.kt

```diff
@@ -125,13 +129,41 @@ fun UserProfileScreen(
                       }
 
                   // Follow/Unfollow Button
-                  val currentUserId = FirebaseAuth.getInstance().currentUser?.uid
                   if (currentUserId != null && currentUserId != userId) {
                     Button(
-                        onClick = { TODO() },
-                        modifier = Modifier.fillMaxWidth().padding(horizontal = 16.dp)) {
-                          Text(if (profile.followers > 0) "Unfollow" else "Follow")
+                        onClick = { viewModel.toggleFollow(currentUserId, userId) },
+                        modifier =
+                            Modifier.fillMaxWidth()
+                                .padding(horizontal = 16.dp)
+                                .testTag("follow_button"),
+                        enabled = followActionState != FollowActionState.Loading) {
+                          when (followActionState) {
+                            FollowActionState.Loading -> {
+                              CircularProgressIndicator(
+                                  modifier = Modifier.size(24.dp),
+                                  color = MaterialTheme.colorScheme.onPrimary)
+                            }
+                            else -> {
+                              when (uiState) {
+                                is ProfileUiState.Success -> {
+                                  val profile = (uiState as ProfileUiState.Success).profile
+                                  Text(
+                                      if (profile.isFollowedByCurrentUser) "Unfollow" else "Follow")
+                                }
+                                else -> Text("Follow")
+                              }
+                            }
+                          }
                         }
+
+                    // Show error if follow action failed
+                    if (followActionState is FollowActionState.Error) {
```

Comment:
Here I think it's better to display a toast that displays the error since errors are commonly displayed this way. But it's just a suggestion if it renders better this way then keep it as it is.

Reply:
Thanks for the suggestions. For the follow button text, I'll keep the current implementation as it maintains good UX with optimistic updates and reasonable defaults. For error display, I prefer keeping the inline error since it provides clear context for the failed action rather than using a transient toast notification. Both the current implementations are functional and follow best practices. These are some optimization tasks, but the thing is my code is working with this current implementation, and i need to change other things to adress this so I will keep in mind ur suggestions for a future opti task in a future PR.